// @antv/f6-plugin/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TimeBar", {
  enumerable: true,
  get: function get() {
    return _timeBar.default;
  }
});
Object.defineProperty(exports, "SnapLine", {
  enumerable: true,
  get: function get() {
    return _snapLine.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _tooltip.default;
  }
});
Object.defineProperty(exports, "Legend", {
  enumerable: true,
  get: function get() {
    return _legend.default;
  }
});
Object.defineProperty(exports, "ZoomSlider", {
  enumerable: true,
  get: function get() {
    return _zoomSlider.default;
  }
});
Object.defineProperty(exports, "EdgeFilterLens", {
  enumerable: true,
  get: function get() {
    return _edgeFilterLens.default;
  }
});
Object.defineProperty(exports, "Fisheye", {
  enumerable: true,
  get: function get() {
    return _fisheye.default;
  }
});
Object.defineProperty(exports, "MiniMap", {
  enumerable: true,
  get: function get() {
    return _minimap.default;
  }
});
Object.defineProperty(exports, "Bundling", {
  enumerable: true,
  get: function get() {
    return _bundling.default;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function get() {
    return _menu.default;
  }
});
exports.default = void 0;

var _timeBar = _interopRequireDefault(require("./time-bar"));

var _snapLine = _interopRequireDefault(require("./snap-line"));

var _tooltip = _interopRequireDefault(require("./tooltip"));

var _legend = _interopRequireDefault(require("./legend"));

var _zoomSlider = _interopRequireDefault(require("./zoom-slider"));

var _edgeFilterLens = _interopRequireDefault(require("./edge-filter-lens"));

var _fisheye = _interopRequireDefault(require("./fisheye"));

var _minimap = _interopRequireDefault(require("./minimap"));

var _bundling = _interopRequireDefault(require("./bundling"));

var _menu = _interopRequireDefault(require("./menu"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var plugin = {
  TimeBar: _timeBar.default,
  SnapLine: _snapLine.default,
  Tooltip: _tooltip.default,
  Legend: _legend.default,
  ZoomSlider: _zoomSlider.default,
  EdgeFilterLens: _edgeFilterLens.default,
  Fisheye: _fisheye.default,
  MiniMap: _minimap.default,
  Bundling: _bundling.default,
  Menu: _menu.default
};
var _default = plugin;
exports.default = _default;
