// @antv/f6-plugin/es/time-bar/utils/const.js
// event
export var SLIDER_PLAY = 'SLIDER_PLAY';
export var SLIDER_PAUSE = 'SLIDER_PAUSE';
export var CONTROL_NEXT = 'CONTROL_NEXT';
export var CONTROL_PREV = 'CONTROL_PREV';
export var RANGE_CHANGE = 'RANGE_CHANGE';
export var SPEED_CHANGE = 'SPEED_CHANGE'; // slider

export var BACKGROUND_STYLE = {
  fill: '#416180',
  opacity: 0.05
};
export var SIMPLE_BACKGROUND_STYLE = {
  fill: '#416180',
  opacity: 0.15,
  radius: 5
};
export var FOREGROUND_STYLE = {
  fill: '#5B8FF9',
  opacity: 0.3,
  cursor: 'grab'
};
export var TREND_SLIDER_HEIHGT = 24;
export var TREND_HANDLER_STYLE = {
  width: 2,
  height: 26,
  fill: '#1890ff',
  stroke: '#1890ff',
  opacity: 1,
  cursor: 'ew-resize',
  // 高亮的颜色
  highLightFill: '#0050b3'
};
export var SIMPLE_HANDLE_STYLE = {
  fill: '#fff',
  stroke: '#1890ff',
  radius: 15,
  opacity: 1,
  lineWidth: 2,
  cursor: 'ew-resize'
}; // sliderTick

export var TEXT_STYLE = {
  textBaseline: 'middle',
  fill: '#000',
  opacity: 0.45,
  fontSize: 12
};
export var TICK_LABEL_STYLE = {
  fill: '#607889',
  opacity: 0.35,
  fontSize: 12,
  width: 50
};
export var TICK_LINE_STYLE = {
  width: 1,
  height: 4,
  fill: '#ccc'
};
export var PROGRESS_TICK = {
  gap: 2,
  tickBoxHeight: 18
};
export var PROGRESS_TICK_LINE_STYLE = {
  height: 4,
  width: 1,
  fill: '#BFBFBF'
};
export var PROGRESS_TICK_TEXT_STYLE = {
  fill: '#8c8c8c',
  fontSize: 10,
  width: 50
};
export var PROGRESS_TICK_SELECTED_STYLE = {
  fill: '#5B8FF9'
};
export var PROGRESS_TICK_UNSELECTED_STYLE = {
  fill: '#e6e8e9'
};
export var SLIDER_START = 0.1;
export var SLIDER_END = 0.9;
export var DEFAULT_CONTROLLER_CONFIG = {
  speed: 1,
  contentWidth: 200,
  height: 60,
  fill: '#fff',
  stroke: '#fff',
  hideTimeTypeController: false,
  preBtnStyle: {
    fill: '#aaa'
  },
  nextBtnStyle: {
    fill: '#aaa'
  },
  playBtnStyle: {
    fill: '#aaa'
  }
};
