// @antv/f6-plugin/lib/time-bar/utils/const.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_CONTROLLER_CONFIG = exports.SLIDER_END = exports.SLIDER_START = exports.PROGRESS_TICK_UNSELECTED_STYLE = exports.PROGRESS_TICK_SELECTED_STYLE = exports.PROGRESS_TICK_TEXT_STYLE = exports.PROGRESS_TICK_LINE_STYLE = exports.PROGRESS_TICK = exports.TICK_LINE_STYLE = exports.TICK_LABEL_STYLE = exports.TEXT_STYLE = exports.SIMPLE_HANDLE_STYLE = exports.TREND_HANDLER_STYLE = exports.TREND_SLIDER_HEIHGT = exports.FOREGROUND_STYLE = exports.SIMPLE_BACKGROUND_STYLE = exports.BACKGROUND_STYLE = exports.SPEED_CHANGE = exports.RANGE_CHANGE = exports.CONTROL_PREV = exports.CONTROL_NEXT = exports.SLIDER_PAUSE = exports.SLIDER_PLAY = void 0;
// event
var SLIDER_PLAY = 'SLIDER_PLAY';
exports.SLIDER_PLAY = SLIDER_PLAY;
var SLIDER_PAUSE = 'SLIDER_PAUSE';
exports.SLIDER_PAUSE = SLIDER_PAUSE;
var CONTROL_NEXT = 'CONTROL_NEXT';
exports.CONTROL_NEXT = CONTROL_NEXT;
var CONTROL_PREV = 'CONTROL_PREV';
exports.CONTROL_PREV = CONTROL_PREV;
var RANGE_CHANGE = 'RANGE_CHANGE';
exports.RANGE_CHANGE = RANGE_CHANGE;
var SPEED_CHANGE = 'SPEED_CHANGE'; // slider

exports.SPEED_CHANGE = SPEED_CHANGE;
var BACKGROUND_STYLE = {
  fill: '#416180',
  opacity: 0.05
};
exports.BACKGROUND_STYLE = BACKGROUND_STYLE;
var SIMPLE_BACKGROUND_STYLE = {
  fill: '#416180',
  opacity: 0.15,
  radius: 5
};
exports.SIMPLE_BACKGROUND_STYLE = SIMPLE_BACKGROUND_STYLE;
var FOREGROUND_STYLE = {
  fill: '#5B8FF9',
  opacity: 0.3,
  cursor: 'grab'
};
exports.FOREGROUND_STYLE = FOREGROUND_STYLE;
var TREND_SLIDER_HEIHGT = 24;
exports.TREND_SLIDER_HEIHGT = TREND_SLIDER_HEIHGT;
var TREND_HANDLER_STYLE = {
  width: 2,
  height: 26,
  fill: '#1890ff',
  stroke: '#1890ff',
  opacity: 1,
  cursor: 'ew-resize',
  // 高亮的颜色
  highLightFill: '#0050b3'
};
exports.TREND_HANDLER_STYLE = TREND_HANDLER_STYLE;
var SIMPLE_HANDLE_STYLE = {
  fill: '#fff',
  stroke: '#1890ff',
  radius: 15,
  opacity: 1,
  lineWidth: 2,
  cursor: 'ew-resize'
}; // sliderTick

exports.SIMPLE_HANDLE_STYLE = SIMPLE_HANDLE_STYLE;
var TEXT_STYLE = {
  textBaseline: 'middle',
  fill: '#000',
  opacity: 0.45,
  fontSize: 12
};
exports.TEXT_STYLE = TEXT_STYLE;
var TICK_LABEL_STYLE = {
  fill: '#607889',
  opacity: 0.35,
  fontSize: 12,
  width: 50
};
exports.TICK_LABEL_STYLE = TICK_LABEL_STYLE;
var TICK_LINE_STYLE = {
  width: 1,
  height: 4,
  fill: '#ccc'
};
exports.TICK_LINE_STYLE = TICK_LINE_STYLE;
var PROGRESS_TICK = {
  gap: 2,
  tickBoxHeight: 18
};
exports.PROGRESS_TICK = PROGRESS_TICK;
var PROGRESS_TICK_LINE_STYLE = {
  height: 4,
  width: 1,
  fill: '#BFBFBF'
};
exports.PROGRESS_TICK_LINE_STYLE = PROGRESS_TICK_LINE_STYLE;
var PROGRESS_TICK_TEXT_STYLE = {
  fill: '#8c8c8c',
  fontSize: 10,
  width: 50
};
exports.PROGRESS_TICK_TEXT_STYLE = PROGRESS_TICK_TEXT_STYLE;
var PROGRESS_TICK_SELECTED_STYLE = {
  fill: '#5B8FF9'
};
exports.PROGRESS_TICK_SELECTED_STYLE = PROGRESS_TICK_SELECTED_STYLE;
var PROGRESS_TICK_UNSELECTED_STYLE = {
  fill: '#e6e8e9'
};
exports.PROGRESS_TICK_UNSELECTED_STYLE = PROGRESS_TICK_UNSELECTED_STYLE;
var SLIDER_START = 0.1;
exports.SLIDER_START = SLIDER_START;
var SLIDER_END = 0.9;
exports.SLIDER_END = SLIDER_END;
var DEFAULT_CONTROLLER_CONFIG = {
  speed: 1,
  contentWidth: 200,
  height: 60,
  fill: '#fff',
  stroke: '#fff',
  hideTimeTypeController: false,
  preBtnStyle: {
    fill: '#aaa'
  },
  nextBtnStyle: {
    fill: '#aaa'
  },
  playBtnStyle: {
    fill: '#aaa'
  }
};
exports.DEFAULT_CONTROLLER_CONFIG = DEFAULT_CONTROLLER_CONFIG;
